packages/legend-query/src/stores/QueryBuilderMilestoningHelper.ts

```diff
@@ -0,0 +1,603 @@
+      derivedPropertyExpressionState.queryBuilderState.graphManagerState.graph,
+      derivedPropertyExpressionState,
+    ) &&
+    guaranteeNonNullable(derivedPropertyExpressionStates[index + 1])
+      .propertyExpression.func.genericType.value.rawType instanceof Class &&
+    derivedPropertyExpressionState.propertyExpression.func.owner
+      ._generatedMilestonedProperties.length !== 0
+  ) {
+    const temporalTarget = getMilestoneTemporalStereotype(
+      guaranteeType(
+        guaranteeNonNullable(derivedPropertyExpressionStates[index + 1])
+          .propertyExpression.func.genericType.value.rawType,
+        Class,
+      ),
+      derivedPropertyExpressionState.queryBuilderState.graphManagerState.graph,
+    );
+    if (temporalTarget) {
+      if (
+        derivedPropertyExpressionStates[index + 1]?.parameterValues &&
+        guaranteeNonNullable(
+          derivedPropertyExpressionStates[index + 1]?.parameterValues,
+        ).length <= idx
+      ) {
+        fillDerivedPropertyArguments(
+          guaranteeNonNullable(derivedPropertyExpressionStates[index + 1]),
+          derivedPropertyExpressionState.queryBuilderState.graphManagerState
+            .graph,
+          true,
+        );
+      }
+    }
+  }
+};
+
+export const getMilestoningDate = (
+  milestoningParameter: ValueSpecification,
+  graph: PureModel,
+): ValueSpecification | undefined => {
+  const type = milestoningParameter.genericType?.value.rawType;
+  const multiplicity = graph.getTypicalMultiplicity(
+    TYPICAL_MULTIPLICITY_TYPE.ONE,
+  );
+  let value;
+  if (milestoningParameter instanceof PrimitiveInstanceValue) {
+    value = milestoningParameter.values[0];
+  }
+  if (type === graph.getPrimitiveType(PRIMITIVE_TYPE.LATESTDATE)) {
+    const parameter = new PrimitiveInstanceValue(
+      GenericTypeExplicitReference.create(
+        new GenericType(graph.getPrimitiveType(PRIMITIVE_TYPE.LATESTDATE)),
+      ),
+      multiplicity,
+    );
+    return parameter;
+  } else if (type === graph.getPrimitiveType(PRIMITIVE_TYPE.STRICTDATE)) {
+    const parameter = new PrimitiveInstanceValue(
+      GenericTypeExplicitReference.create(
+        new GenericType(graph.getPrimitiveType(PRIMITIVE_TYPE.STRICTDATE)),
+      ),
+      multiplicity,
+    );
+    parameter.addValue(value);
+    return parameter;
+  } else if (type === graph.getPrimitiveType(PRIMITIVE_TYPE.DATETIME)) {
+    const parameter = new PrimitiveInstanceValue(
+      GenericTypeExplicitReference.create(
+        new GenericType(graph.getPrimitiveType(PRIMITIVE_TYPE.DATETIME)),
+      ),
+      multiplicity,
+    );
+    parameter.addValue(value);
+    return parameter;
+  } else if (milestoningParameter instanceof VariableExpression) {
+    const parameter = new VariableExpression(
+      milestoningParameter.name,
+      multiplicity,
+    );
+    return parameter;
+  }
+  return undefined;
+};
+
+export const getPropertyExpressionState = (
```

Comment:
can we change the name of this to what exactly the function is doing to the property expression state

Reply:
Seems like we may be able to do something similar to what we do in model to model mapping. We `decorate` the derived property when the date is propogated and we clean after the editor is closed.